### PR TITLE
feat(trace): http api for trace

### PR DESCRIPTION
**Pull request by zhongwencool**

### CODE REVIEW

File: lib-ce/emqx_modules/src/emqx_mod_trace.erl [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+    {NeedRunning, AllStarted} = start_trace(Running, Started),
+    NeedStop = AllStarted -- NeedRunning,
+    stop_trace(NeedStop, Started),
+    clean_stale_trace_files(),
+    NextTime = find_closest_time(Traces, Now),
+    emqx_misc:start_timer(NextTime, update_trace).
+
+stop_all_trace_handler() ->
+    lists:foreach(fun({Type, Value, Name}) ->
+        _ = emqx_tracer:stop_trace({Type, Value, Name}) end
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_trace.erl:258]: emqx_trace outputs `string()`s, but here we pass in `binary()` to `emqx_trace` module.
maybe unify the behaviour.

**Review by zmstone — COMMENTED** [review]
just a few comments from a quick glance

File: src/emqx_tracer.erl [hunk 2]
```diff
@@ -24,11 +24,18 @@
...
+        , stop_trace/2
         ]).
 
--type(trace_who() :: {clientid | topic, binary()}).
+-ifdef(TEST).
+-export([is_match/5]).
+-endif.
+
+-type(trace_who_with_context() :: {clientid | topic, emqx_types:clientid() | emqx_types:topic(), binary(), [atom()]}).
+-type(trace_who() :: {clientid | topic, emqx_types:clientid() | emqx_types:topic(), binary()}).
```
Comment (zmstone) [inline src/emqx_tracer.erl:38]: maybe make this an `opaque()`
i.e. no other module should have to know this is a 3-tuple.

File: src/emqx_tracer.erl [hunk 3]
```diff
@@ -59,28 +81,47 @@
...
+trace_subscribe(<<"$SYS/", _/binary>>, _SubId, _SubOpts) -> ignore;
+trace_subscribe(Topic, SubId, SubOpts) ->
+    emqx_logger:info(#{topic => Topic,
+        mfa => {?MODULE, ?FUNCTION_NAME, ?FUNCTION_ARITY}},
+        "~s SUBSCRIBE ~s: Options: ~0p", [SubId, Topic, SubOpts]).
+
+trace_unsubscribe(<<"$SYS/", _/binary>>, _SubOpts) -> ignore;
+trace_unsubscribe(Topic, SubOpts) ->
+    emqx_logger:info(#{topic => Topic, mfa => {?MODULE, ?FUNCTION_NAME, ?FUNCTION_ARITY}},
+        "~s UNSUBSCRIBE ~s: Options: ~0p",
```
Comment (zmstone) [inline src/emqx_tracer.erl:105]:
```suggestion
        "~ts UNSUBSCRIBE ~ts: Options: ~0p",
```

File: src/emqx_tracer.erl [hunk 4]
```diff
@@ -59,28 +81,47 @@
...
         when is_binary(From); is_atom(From) ->
     emqx_logger:info(#{topic => Topic,
                        mfa => {?MODULE, ?FUNCTION_NAME, ?FUNCTION_ARITY} },
                      "PUBLISH to ~s: ~0p", [Topic, Payload]).
 
+trace_subscribe(<<"$SYS/", _/binary>>, _SubId, _SubOpts) -> ignore;
+trace_subscribe(Topic, SubId, SubOpts) ->
+    emqx_logger:info(#{topic => Topic,
+        mfa => {?MODULE, ?FUNCTION_NAME, ?FUNCTION_ARITY}},
+        "~s SUBSCRIBE ~s: Options: ~0p", [SubId, Topic, SubOpts]).
```
Comment (zmstone) [inline src/emqx_tracer.erl:100]:
```suggestion
        "~ts SUBSCRIBE ~ts: Options: ~0p", [SubId, Topic, SubOpts]).
```

File: lib-ce/emqx_modules/src/emqx_mod_trace_api.erl [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                [ZipName | Acc];
+            {error, Node, Reason} ->
+                ?LOG(error, "download trace log error:~p", [{Node, TraceLog, Reason}]),
+                Acc
+        end
+                end, [], TraceFiles).
+
+collect_trace_file(TraceLog) ->
+    Nodes = ekka_mnesia:running_nodes(),
+    {File, BadNodes} = rpc:multicall(Nodes, emqx_mod_trace, trace_file, [TraceLog], 60000),
```
Comment (zmstone) [inline lib-ce/emqx_modules/src/emqx_mod_trace_api.erl:144]:
```suggestion
    {Files, BadNodes} = rpc:multicall(Nodes, emqx_mod_trace, trace_file, [TraceLog], 60000),
```

File: apps/emqx_management/src/emqx_mgmt_cli.erl [hunk 6]
```diff
@@ -406,15 +413,16 @@ log(_) ->
...
+                        "Set log level of a log handler"}]).
 
 %%--------------------------------------------------------------------
 %% @doc Trace Command
 
 trace(["list"]) ->
-    lists:foreach(fun({{Who, Name}, {Level, LogFile}}) ->
-                emqx_ctl:print("Trace(~s=~s, level=~s, destination=~p)~n", [Who, Name, Level, LogFile])
-            end, emqx_tracer:lookup_traces());
+    lists:foreach(fun({{Who, Name, _}, {Level, LogFile}}) ->
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_cli.erl:423]: should be a map() now ?

**Comment by zmstone** [comment]
the cli command (and maybe http API too?) defaults to write the trace file at EMQ X root dir.
should use the log dir instead.

File: lib-ce/emqx_modules/src/emqx_mod_trace.erl [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+    ets:match_object(?TRACE, #?TRACE{_ = '_'}).
+
+-spec list(boolean()) -> [tuple()].
+list(Enable) ->
+    ets:match_object(?TRACE, #?TRACE{enable = Enable, _ = '_'}).
+
+-spec create([{Key :: binary(), Value :: binary()}]) ->
+    ok | {error, {duplicate_condition, iodata()} | {already_existed, iodata()} | iodata()}.
+create(Trace) ->
+    case mnesia:table_info(?TRACE, size) < ?MAX_SIZE of
```
[the remark this answers is not in the document]
Comment (zhongwencool) [reply lib-ce/emqx_modules/src/emqx_mod_trace.erl:116]: when size =:= Max_Size  we should return an error.

**Review by zmstone — APPROVED** [review]

**Pull request merged**